numpy/core/src/multiarray/dlpack.c

```diff
@@ -14,10 +14,21 @@
 static void
 array_dlpack_deleter(DLManagedTensor *self)
 {
+    // Leak the pyobj if not initialized.  This can happen if we are running
+    // exit handlers that are destructing c++ objects with residual (owned)
+    // PyObjects stored in them after the Python runtime has already been
+    // terminated.
+    if (!Py_IsInitialized())
+        return;
```

Comment:
Please add brackets (slight preference for `/*` style comments also).

Curious, is this really expected or more of a hack for programs which don't ensure cleanup has a well defined order?

Reply:
> Curious, is this really expected or more of a hack for programs which don't ensure cleanup has a well defined order?

This is a real problem in PyTorch. In particular, we do have quite a few long-lived c++ objects. And because C++ cleanup order is not really guaranteed, it is possible for such a long lived c++ object to hold onto a pyobject. We had quite a few deadlocks because of it https://github.com/pytorch/pytorch/pull/73961 and https://github.com/pytorch/pytorch/pull/57488 are example of cleanups we did there.

Reply:
I can remove this if you prefer, but given that our Tensor object is potentially one such c++ owned object and it will call this function on destruction. I would prefer to keep it here.